File: apps/movie-app/movie-catalog-service/movie_catalog_service.py
import requests, os
from flask import Flask, jsonify

app = Flask(__name__)


# Endpoint URLs for Movie Info and Movie Ratings services
# Access the environment variables
movie_info_url = os.environ.get('MOVIE_INFO_URL', 'http://default-movie-info-url')
movie_ratings_url = os.environ.get('MOVIE_RATINGS_URL', 'http://default-movie-ratings-url')


# Function to get movie details
@app.route('/api/movie/<int:movie_id>', methods=['GET'])
def get_movie_details(movie_id):
    # Make a request to the Movie Info service
    response_info = requests.get(f"{movie_info_url}/{movie_id}")
    movie_info = response_info.json()

    # Make a request to the Movie Ratings service
    response_ratings = requests.get(f"{movie_ratings_url}/{movie_id}")
    movie_ratings = response_ratings.json()

    # Combine movie info and ratings
    movie_details = {
        "movie_info": movie_info,
        "movie_ratings": movie_ratings,
    }

    return movie_details

if __name__ == '__main__':
    app.run(debug=True, host='0.0.0.0', port=5000)

File: apps/movie-app/movie-catalog-service/movie_catalog_service.go
package main

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strconv"
)

var movieInfoURL = getEnv("MOVIE_INFO_URL", "http://default-movie-info-url")
var movieRatingsURL = getEnv("MOVIE_RATINGS_URL", "http://default-movie-ratings-url")

func getEnv(key, defaultValue string) string {
	value := os.Getenv(key)
	if value == "" {
		return defaultValue
	}
	return value
}

func getMovieDetails(w http.ResponseWriter, r *http.Request) {
	movieIDStr := r.URL.Query().Get("movie_id")
	movieID, err := strconv.Atoi(movieIDStr)
	if err != nil {
		http.Error(w, "Invalid movie ID", http.StatusBadRequest)
		return
	}

	// Make a request to the Movie Info service
	responseInfo, err := http.Get(fmt.Sprintf("%s/%d", movieInfoURL, movieID))
	if err != nil {
		http.Error(w, "Failed to fetch movie info", http.StatusInternalServerError)
		return
	}
	defer responseInfo.Body.Close()

	var movieInfo map[string]interface{}
	if err := json.NewDecoder(responseInfo.Body).Decode(&movieInfo); err != nil {
		http.Error(w, "Failed to decode movie info response", http.StatusInternalServerError)
		return
	}

	// Make a request to the Movie Ratings service
	responseRatings, err := http.Get(fmt.Sprintf("%s/%d", movieRatingsURL, movieID))
	if err != nil {
		http.Error(w, "Failed to fetch movie ratings", http.StatusInternalServerError)
		return
	}
	defer responseRatings.Body.Close()

	var movieRatings map[string]interface{}
	if err := json.NewDecoder(responseRatings.Body).Decode(&movieRatings); err != nil {
		http.Error(w, "Failed to decode movie ratings response", http.StatusInternalServerError)
		return
	}

	// Combine movie info and ratings
	movieDetails := map[string]interface{}{
		"movie_info":    movieInfo,
		"movie_ratings": movieRatings,
	}

	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(movieDetails)
}

func main() {
	http.HandleFunc("/api/movie", getMovieDetails)

	// Start the HTTP server
	fmt.Println("Server listening on :5000")
	if err := http.ListenAndServe(":5000", nil); err != nil {
		fmt.Printf("Server error: %s\n", err)
	}
}
